go/vt/mysqlctl/backupengine.go

```diff
@@ -138,23 +142,30 @@ type BackupManifest struct {
 // FindBackupToRestore returns a selected candidate backup to be restored.
 // It returns the most recent backup that is complete, meaning it has a valid
 // MANIFEST file.
-func FindBackupToRestore(ctx context.Context, cnf *Mycnf, mysqld MysqlDaemon, logger logutil.Logger, dir string, bhs []backupstorage.BackupHandle) (backupstorage.BackupHandle, error) {
+func FindBackupToRestore(ctx context.Context, cnf *Mycnf, mysqld MysqlDaemon, logger logutil.Logger, dir string, bhs []backupstorage.BackupHandle, snapshotTime time.Time) (backupstorage.BackupHandle, error) {
 	var bh backupstorage.BackupHandle
 	var index int
+	unixZeroTime := time.Time{}
 
 	for index = len(bhs) - 1; index >= 0; index-- {
 		bh = bhs[index]
 		// Check that the backup MANIFEST exists and can be successfully decoded.
-		_, err := GetBackupManifest(ctx, bh)
+		bm, err := GetBackupManifest(ctx, bh)
 		if err != nil {
 			log.Warningf("Possibly incomplete backup %v in directory %v on BackupStorage: can't read MANIFEST: %v)", bh.Name(), dir, err)
 			continue
 		}
 
-		logger.Infof("Restore: found backup %v %v to restore", bh.Directory(), bh.Name())
-		break
+		backupTime, _ := time.Parse(time.RFC3339, bm.BackupTime)
+		if snapshotTime.Equal(unixZeroTime) /* uninitialized or not snapshot */ || bm.BackupTime == "" /* restoring an older backup where MANIFEST does not have backupTime */ || backupTime.Before(snapshotTime) {
```

Comment:
If a non-zero `snapshotTime` is requested, but we can't enforce it because the latest backup doesn't have BackupTime yet, shouldn't we return an error to indicate that we can't do what was asked?

Reply:
we do have an error later on which covers this condition. do you think we should have a more specific error?

```	if index < 0 {
		if !snapshotTime.IsZero() {
			log.Errorf("No valid backup found before time %v", snapshotTime.Format("2006-01-02.150405"))
		}
		// There is at least one attempted backup, but none could be read.
		// This implies there is data we ought to have, so it's not safe to start
		// up empty.
		return nil, ErrNoCompleteBackup
	}
```

Reply:
It looks like `ErrNoCompleteBackup` will only be returned if we *don't* break here. Currently, if a non-zero snapshotTime is requested, but `bm.BackupTime == ""`, we will break. That means `index` is still `>= 0` so we won't enter the `if` block below.

Reply:
Good catch. I've changed it to not break on that condition. IOW, if there's no backup with a BackupTime != "" then we can't recover a snapshot. I have left a TODO in there until we decide whether we should infer the BackupTime from the name (which contains a timestamp)